services/inputhost/inputhost.go

```diff
@@ -641,6 +647,105 @@ func (h *InputHost) DestinationsUpdated(ctx thrift.Context, request *admin.Desti
+		h.pathMutex.RLock()
+		pathCache, ok := h.pathCache[destUUID]
+		if ok {
+			pathCache.Lock()
+			pathCache.prepareForUnload()
+			go pathCache.unload()
+			pathCache.Unlock()
+		} else {
+			h.logger.WithField(common.TagDst, common.FmtDst(destUUID)).
+				Error("destination is not cached at all")
+			err = ErrDstNotLoaded
+			h.m3Client.IncCounter(metrics.UnloadDestinationsScope, metrics.InputhostFailures)
+		}
+		h.pathMutex.RUnlock()
+	}
+
+	return err
+}
+
+// ListLoadedDestinations is the API used to list all the loaded destinations in memory
+func (h *InputHost) ListLoadedDestinations(ctx thrift.Context) (result *admin.ListDestinationsResult_, err error) {
+	defer atomic.AddInt32(&h.loadShutdownRef, -1)
+	sw := h.m3Client.StartTimer(metrics.ListLoadedDestinationsScope, metrics.InputhostLatencyTimer)
+	defer sw.Stop()
+	h.m3Client.IncCounter(metrics.ListLoadedDestinationsScope, metrics.InputhostRequests)
+	// If we are already shutting down, no need to do anything here
+	if atomic.AddInt32(&h.loadShutdownRef, 1) <= 0 {
+		h.logger.Error("inputHost already shutdown")
+		h.m3Client.IncCounter(metrics.ListLoadedDestinationsScope, metrics.InputhostFailures)
+		return nil, ErrHostShutdown
+	}
+
+	result = admin.NewListDestinationsResult_()
+	result.Dests = make([]*admin.Destinations, 0)
+	h.pathMutex.RLock()
+	for destUUID, pathCache := range h.pathCache {
+		destRes := admin.NewDestinations()
+		destRes.DestUUID = common.StringPtr(destUUID)
+		destRes.DestPath = common.StringPtr(pathCache.destinationPath)
+
+		result.Dests = append(result.Dests, destRes)
+	}
+	h.pathMutex.RUnlock()
+
+	return result, err
+}
+
+// ReadDestState is the API used to read the state of the destination which is loaded on thi inputhost
+func (h *InputHost) ReadDestState(ctx thrift.Context, request *admin.ReadDestinationStateRequest) (result *admin.ReadDestinationStateResult_, err error) {
+	defer atomic.AddInt32(&h.loadShutdownRef, -1)
+	sw := h.m3Client.StartTimer(metrics.ReadDestStateScope, metrics.InputhostLatencyTimer)
+	defer sw.Stop()
+	h.m3Client.IncCounter(metrics.ReadDestStateScope, metrics.InputhostRequests)
+	// If we are already shutting down, no need to do anything here
+	if atomic.AddInt32(&h.loadShutdownRef, 1) <= 0 {
+		h.logger.Error("inputHost already shutdown")
+		h.m3Client.IncCounter(metrics.ReadDestStateScope, metrics.InputhostFailures)
+		return nil, ErrHostShutdown
+	}
+
+	result = admin.NewReadDestinationStateResult_()
+
+	result.InputHostUUID = common.StringPtr(h.GetHostUUID())
+	result.DestState = make([]*admin.DestinationState, 0)
+
+	// Now populate all destination state
+	h.pathMutex.RLock()
+	for _, destUUID := range request.DestUUIDs {
+		pathCache, ok := h.pathCache[destUUID]
+		if ok {
+			destState := pathCache.getState()
+			result.DestState = append(result.DestState, destState)
+		} else {
+			h.logger.WithField(common.TagDst, common.FmtDst(destUUID)).
+				Error("destination is not cached at all")
+			err = ErrDstNotLoaded
```

Comment:
seems more like a warning to me since this can happen in normal situation. Also since you're continuing with other destination when you see an error, increasing the counter below could cause more than 1 failures for each request